### ISSUE TITLE
How to prevent advancing the HL consumer offset if error occurs? ie. Block at err offset.

### ISSUE THREAD

**Issue by RobvH** [confluentinc/confluent-kafka-dotnet#903]
Have based a current high-level consumer on the example consumer.cs here. That example shows how to commit offsets, due to the time expense, every few messages. `consumer.Consume` is called in the loop and every 5 messages the offset is committed. 

This means the offset it being advanced somewhere internally at least to the consumer, otherwise Consume() would keep returning the same message, right? What I want to do is be able to control when Consume() gets the next message or stays blocked at the current one and gives it again.

The rational is straightforward -- I have a topic of events that must be processed and persisted. If persistence fails because the database is too busy or aws is flaky that moment, or for whatever reason I decide, I want to sleep, loop again over the Consume call again and get the same (incompletely processed message). I want to continue getting the message at that offset indefinitely until I deem it a success.

In an older PHP HL client using rdkafka, a teammate did this by unsubscribing then re-subscribing to the consumer to the topic inside a catch block so that anytime there were errors worth blocking for, the consumer would throw and the catch would basically clear it's memory that it had read that message. Of course, commits manually if everything is a success. This constant unsubscribing seems potentially problematic to me. I really just want my client to catch whatever exception we have, wait 15 seconds, then loop again; having the Consume() call again return that message to try over. 

Is the right approach set an inner try-catch and within the catch to call 
```
consumer.Seek(new TopicPartitionOffset(consumeResult.TopicPartition, -1));
```
and then let it loop? Otherwise, letting it commit every 5 offsets or so? Is there a better approach?

Confluent.Kafka 1.0.0-beta2, dotnet core 2.2

**Comment by mhowlett** [comment]
The pattern you describe isn't usual. I'd suggest just consuming a message, then processing in a retry loop, and committing the offset after the processing is done. It sounds like you need to produce to your external system synchronously (which is unfortunate). Given that it's probably not a big deal to use `Commit` to commit the offset, but you could also consider setting `EnableAutoOffsetStore` to false and using `StoreOffsets` instead, which will give you async commits whilst still provide at-least once semantics.

**Comment by RobvH** [comment]
Great response, thanks! Yes, at the partition level we have to remain FIFO, but luckily the key we can partition on gives us very wide concurrency in our data set. There is a more complex circuit-breaker logic for deciding when to seek back or not in the production code, I just wanted to present the essential algorithm here and validate the use of `Seek()`. 

The only other example code I'd seen that was using RdKafka did this by unsubscribing and re-subscribing the consumer as the PHP extension doesn't expose Seek. Glad we do on this wrapper.

Basically, while we're failing on recoverable exceptions (connectivity) we do a growing back off and seek back. When we encounter unrecoverable exceptions we log profusely and carry on with the stream. We just want to make sure our queues handle network issues resiliently.

Thanks again for your feedback!
